File: src/lib/tracker/motor.ts
import { clamp } from "./math.ts";

export interface MotorAxis {
  positionDeg: number;
  currentVel: number;
  targetVel: number;
  enabled: boolean;
}

export function createMotor(positionDeg = 0): MotorAxis {
  return {
    positionDeg,
    currentVel: 0,
    targetVel: 0,
    enabled: false,
  };
}

/**
 * First-order acceleration-limited velocity ramp, then integrate position.
 * Speeds are in deg/s. Never jumps to max speed.
 */
export function stepMotor(
  axis: MotorAxis,
  dt: number,
  maxSpeed: number,
  acceleration: number,
  limitDeg: number,
): MotorAxis {
  if (!axis.enabled) {
    return {
      ...axis,
      currentVel: 0,
      targetVel: 0,
    };
  }

  const target = clamp(axis.targetVel, -maxSpeed, maxSpeed);
  const maxDv = Math.max(0, acceleration) * dt;
  const dv = clamp(target - axis.currentVel, -maxDv, maxDv);
  let vel = axis.currentVel + dv;
  vel = clamp(vel, -maxSpeed, maxSpeed);

  let pos = axis.positionDeg + vel * dt;
  if (pos > limitDeg) {
    pos = limitDeg;
    vel = Math.min(0, vel);
  } else if (pos < -limitDeg) {
    pos = -limitDeg;
    vel = Math.max(0, vel);
  }

  return {
    ...axis,
    positionDeg: pos,
    currentVel: vel,
    targetVel: target,
  };
}

export function disableMotor(axis: MotorAxis): MotorAxis {
  return { ...axis, enabled: false, currentVel: 0, targetVel: 0 };
}

export function enableMotor(axis: MotorAxis): MotorAxis {
  return { ...axis, enabled: true };
}

File: firmware/Drivers/motor.c
#include "motor.h"

static float clampf(float v, float lo, float hi) {
    if (v < lo) return lo;
    if (v > hi) return hi;
    return v;
}

void motor_step(motor_axis_t *ax, float dt) {
    if (!ax->enabled) {
        ax->current_vel = 0.f;
        ax->target_vel = 0.f;
        return;
    }
    float target = clampf(ax->target_vel, -ax->max_speed, ax->max_speed);
    float max_dv = ax->acceleration * dt;
    float dv = target - ax->current_vel;
    if (dv > max_dv) dv = max_dv;
    if (dv < -max_dv) dv = -max_dv;
    ax->current_vel = clampf(ax->current_vel + dv, -ax->max_speed, ax->max_speed);
    ax->position_deg += ax->current_vel * dt;
    if (ax->position_deg > ax->limit_deg) {
        ax->position_deg = ax->limit_deg;
        if (ax->current_vel > 0) ax->current_vel = 0;
    } else if (ax->position_deg < -ax->limit_deg) {
        ax->position_deg = -ax->limit_deg;
        if (ax->current_vel < 0) ax->current_vel = 0;
    }
}
